src/realm/cluster_tree.cpp

```diff
@@ -1126,99 +1127,93 @@ bool ClusterTree::is_string_enum_type(ColKey::Idx col_ndx) const
 void ClusterTree::remove_all_links(CascadeState& state)
 {
     Allocator& alloc = get_alloc();
+    auto origin_table = get_owning_table();
     // This function will add objects that should be deleted to 'state'
-    auto func = [this, &state, &alloc](const Cluster* cluster) {
+    auto func = [&](const Cluster* cluster) {
         auto remove_link_from_column = [&](ColKey col_key) {
             // Prevent making changes to table that is going to be removed anyway
             // Furthermore it is a prerequisite for using 'traverse' that the tree
             // is not modified
-            if (get_owning_table()->links_to_self(col_key)) {
+            if (origin_table->links_to_self(col_key)) {
                 return IteratorControl::AdvanceToNext;
             }
             auto col_type = col_key.get_type();
-            if (col_key.is_list() || col_key.is_set()) {
-                if (col_type == col_type_LinkList)
-                    col_type = col_type_Link;
-                if (col_type == col_type_Link) {
-                    ArrayInteger values(alloc);
-                    cluster->init_leaf(col_key, &values);
-                    size_t sz = values.size();
-                    BPlusTree<ObjKey> links(alloc);
-                    for (size_t i = 0; i < sz; i++) {
-                        if (ref_type ref = values.get_as_ref(i)) {
-                            links.init_from_ref(ref);
-                            if (links.size() > 0) {
-                                cluster->remove_backlinks(cluster->get_real_key(i), col_key, links.get_all(), state);
+            if (col_type == col_type_LinkList)
+                col_type = col_type_Link;
+            auto nesting_levels = origin_table->get_nesting_levels(col_key);
+            if (col_key.is_collection()) {
+                ArrayInteger values(alloc);
+                cluster->init_leaf(col_key, &values);
+                size_t sz = values.size();
+
+                for (size_t i = 0; i < sz; i++) {
+                    if (ref_type ref = values.get_as_ref(i)) {
+                        ObjKey origin_key = cluster->get_real_key(i);
+                        if (nesting_levels > 0) {
+                            if (col_type == col_type_Link) {
+                                DummyParent parent(origin_table->m_own_ref, ref);
+                                auto list = CollectionList::create(parent, col_key);
+                                std::vector<ObjKey> keys;
+                                list->get_all_keys(nesting_levels - 1, keys);
+                                cluster->do_remove_backlinks(origin_key, col_key, keys, state);
```

Comment:
Have we thought about the performance implications here? Also, there is a chance that for deep nested collections, we may run into memory problems due to stack overflow issues.

Reply:
We are very aware that this is expensive. We will at some point introduce a maximum nesting level of 100, so should not be a problem.